src/Scriban/Syntax/ScriptVisitor.cs

```diff
@@ -0,0 +1,767 @@
+
+        protected virtual ScriptNode VisitElseStatement(ScriptElseStatement node)
+        {
+            var newBody = (ScriptBlockStatement)Visit(node.Body);
+            var newElse = (ScriptConditionStatement)Visit(node.Else);
+
+            if (newBody == node.Body &&
+                newElse == node.Else)
+            {
+                return node;
+            }
+
+            return new ScriptElseStatement
+            {
+                Body = newBody,
+                Else = newElse
+            }.WithTriviaAndSpanFrom(node);
+        }
+
+        protected virtual ScriptNode VisitForStatement(ScriptForStatement node)
+        {
+            var newVariable = (ScriptExpression)Visit(node.Variable);
+            var newIterator = (ScriptExpression)Visit(node.Iterator);
+            var newNamedArguments = VisitAll(node.NamedArguments);
+            var newBody = (ScriptBlockStatement)Visit(node.Body);
+
+            if (newVariable == node.Variable &&
+                newIterator == node.Iterator &&
+                newNamedArguments == node.NamedArguments &&
+                newBody == node.Body)
+            {
+                return node;
+            }
+
+            return new ScriptForStatement
+            {
+                Variable = newVariable,
+                Iterator = newIterator,
+                NamedArguments = newNamedArguments,
+                Body = newBody
+            }.WithTriviaAndSpanFrom(node);
+        }
+
+        protected virtual ScriptNode VisitIfStatement(ScriptIfStatement node)
+        {
+            var newCondition = (ScriptExpression)Visit(node.Condition);
+            var newThen = (ScriptBlockStatement)Visit(node.Then);
+            var newElse = (ScriptConditionStatement)Visit(node.Else);
+
+            if (newCondition == node.Condition &&
+                newThen == node.Then &&
+                newElse == node.Else)
+            {
+                return node;
+            }
+
+            return new ScriptIfStatement
+            {
+                Condition = newCondition,
+                Then = newThen,
+                Else = newElse,
+                InvertCondition = node.InvertCondition,
+                IsElseIf = node.IsElseIf
+            }.WithTriviaAndSpanFrom(node);
+        }
+
+        protected virtual ScriptNode VisitWhenStatement(ScriptWhenStatement node)
+        {
+            var newBody = (ScriptBlockStatement)Visit(node.Body);
+            var newNext = (ScriptConditionStatement)Visit(node.Next);
+            var newValues = VisitAll(node.Values);
+
+            if (newBody == node.Body &&
+                newNext == node.Next &&
+                newValues == node.Values)
+            {
+                return node;
+            }
+
+            var newStatement = new ScriptWhenStatement
+            {
+                Body = newBody,
+                Next = newNext
+            }.WithTriviaAndSpanFrom(node);
+
+            newStatement.Values.AddRange(newValues);
```

Comment:
Any reason to make `Values` read-only? The code would be more efficient (and clearer) if it weren't. The nodes aren't immutable anyway, and collections are exposed as `List<T>`. Same is true of other node types.

(I assume the main benefit of making it read-only is that you can be sure it's never null)